// barista/lib/resource.js
// Generated by CoffeeScript 1.6.2
(function() {
  var Resource, inflection;

  inflection = require('inflection');

  exports.Resource = Resource = (function() {
    function Resource(base, controller) {
      var slug;

      slug = inflection.underscore(inflection.pluralize(controller));
      this.routes = [base.get("/" + slug + "(.:format)").to("" + controller + ".index"), base.post("/" + slug + "(.:format)").to("" + controller + ".create"), base.get("/" + slug + "/add(.:format)").to("" + controller + ".add"), base.get("/" + slug + "/:id(.:format)").to("" + controller + ".show"), base.get("/" + slug + "/:id/edit(.:format)").to("" + controller + ".edit"), base.put("/" + slug + "/:id(.:format)").to("" + controller + ".update"), base.del("/" + slug + "/:id(.:format)").to("" + controller + ".destroy")];
      this.collection_route = this.routes[0];
      this.member_route = this.routes[3];
      this.collection_route.collection = true;
      this.member_route.member = true;
      this;
    }

    Resource.prototype.collection = function(cb) {
      this.collection_route.nest(cb);
      return this;
    };

    Resource.prototype.member = function(cb) {
      this.member_route.nest(cb);
      return this;
    };

    Resource.prototype.nest = function(cb) {
      this.member_route.nest(cb);
      this.collection_route.nest(cb);
      return this;
    };

    return Resource;

  })();

}).call(this);
